### PR TITLE
Improved performance for export nodes

### PR DESCRIPTION
**Pull request by IIpragmaII**

### CODE REVIEW

File: src/main/java/com/lothrazar/storagenetwork/util/RequestBatch.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+
+public class RequestBatch extends Batch<Request> {
+    public void extractStacks(IConnectableLink providerStorage, Integer slot, Item item) {
+        List<Request> requests = get(item);
+        List<Request> remainingRequests = new ArrayList<Request>();
+        for (Request request : requests) {
+            if (!request.insertStack(providerStorage, slot)) {
+                remainingRequests.add(request);
```
Comment (IIpragmaII) [inline src/main/java/com/lothrazar/storagenetwork/util/RequestBatch.java:17]: There could be a better way to track the remaining requests, than creating a new list for them

File: src/main/java/com/lothrazar/storagenetwork/util/Request.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import com.lothrazar.storagenetwork.api.IConnectableLink;
+
+import net.minecraft.world.item.ItemStack;
+
+public class Request {
+    private Integer count = 0;
+    private IConnectableItemAutoIO storage;
+
+    public Request(IConnectableItemAutoIO storage) {
+        this.count = storage.getTransferRate();
```
Comment (IIpragmaII) [inline src/main/java/com/lothrazar/storagenetwork/util/Request.java:13]: Maybe put the calculation for stock mode here.

File: src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java [hunk 3]
```diff
@@ -178,12 +185,15 @@ private void updateProcess() {
    */
   private void updateExports() {
     Set<IConnectable> conSet = nw.getConnectables();
+    RequestBatch requstBatch = new RequestBatch();
+
     for (IConnectable connectable : conSet) {
       if (connectable == null || connectable.getPos() == null) {
-        //        StorageNetwork.log("null connectable or pos : updateExports() ");
```
Comment (IIpragmaII) [inline src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java:183]: Whitespace changes should be reverted.

File: src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java [hunk 4]
```diff
@@ -178,12 +185,15 @@ private void updateProcess() {
    */
   private void updateExports() {
     Set<IConnectable> conSet = nw.getConnectables();
+    RequestBatch requstBatch = new RequestBatch();
```
Comment (VasurTrekkson) [inline src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java]:
```suggestion
    RequestBatch requestBatch = new RequestBatch();
```

File: src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java [hunk 5]
```diff
@@ -199,66 +209,49 @@ private void updateExports() {
...
-        // Alright, some items got moved in the simulation. Let's do it for real this time.
-        ItemStack realExtractedStack = request(new ItemStackMatcher(requestedStack, false, true), requestedStack.getCount(), false);
-        if (realExtractedStack.isEmpty()) {
+        if (matcher.getStack().isEmpty() || request.getCount() == 0) {
+          // either the thing is empty or we are requesting none
           continue;
         }
-        storage.insertStack(realExtractedStack, false);
-        break;
+        requstBatch.put(matcher.getStack().getItem(), request);
```
Comment (VasurTrekkson) [inline src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java]:
```suggestion
        requestBatch.put(matcher.getStack().getItem(), request);
```

File: src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java [hunk 6]
```diff
@@ -199,66 +209,49 @@ private void updateExports() {
...
       }
     }
+    executeRequestBatch(requstBatch);
```
Comment (VasurTrekkson) [inline src/main/java/com/lothrazar/storagenetwork/block/main/TileMain.java]:
```suggestion
    executeRequestBatch(requestBatch);
```

File: src/main/java/com/lothrazar/storagenetwork/util/Request.java (hunk 2 above)
Comment (IIpragmaII) [reply src/main/java/com/lothrazar/storagenetwork/util/Request.java:13]: You would need to hand the current level and the matcher to the request. I don't think thats worth it. Makes it more confusing, than easier to read.

File: src/main/java/com/lothrazar/storagenetwork/util/RequestBatch.java (hunk 1 above)
Comment (IIpragmaII) [reply src/main/java/com/lothrazar/storagenetwork/util/RequestBatch.java:17]: Gonna keep it this way for the time beeing. editing the list while in the loop throws an error.

**Review by VasurTrekkson — APPROVED** [review]
Ok

**Pull request merged**